File: cognition_fabric/fabric.py
import sqlite3
import json
from typing import List, Callable, Optional
from models import Insight
from guardrail import AutonomousGuardrail

class CognitionFabric:
    """
    The Cognition Fabric Layer: A shared memory mechanism backed by SQLite 
    (for ledger durability) and an in-memory event bus (for the Accelerator mechanism).
    """
    def __init__(self, db_path: str = ":memory:"):
        self.db_path = db_path
        self.conn = sqlite3.connect(self.db_path, check_same_thread=False)
        self.guardrail = AutonomousGuardrail(min_confidence=0.80)
        self.subscribers: List[Callable[[Insight], None]] = []
        self._init_db()

    def _init_db(self):
        cursor = self.conn.cursor()
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS insights (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                problem_signature TEXT UNIQUE,
                solution TEXT,
                confidence_score REAL,
                source_agent_id TEXT,
                metadata TEXT,
                timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
            )
        """)
        self.conn.commit()

    def subscribe(self, callback: Callable[[Insight], None]):
        """Accelerator Mechanism: Register an agent to receive real-time updates."""
        self.subscribers.append(callback)

    def publish_insight(self, insight: Insight) -> bool:
        """
        Accepts a proposed insight, validates it via Guardrails, and stores it in the Fabric.
        """
        print(f"\n[Fabric] 📩 Received proposed insight from {insight.source_agent_id} for '{insight.problem_signature}'")
        
        # 1. Guardrail Validation
        is_valid, reason = self.guardrail.validate(insight)
        if not is_valid:
            print(f"[Guardrail BLOCK] 🛑 {reason}")
            return False

        # 2. Consistency & Storage
        cursor = self.conn.cursor()
        try:
            # Upsert logic to manage Fabric Consistency at Scale (Semantic Locking/Scoped Consistency)
            cursor.execute("SELECT confidence_score FROM insights WHERE problem_signature = ?", (insight.problem_signature,))
            existing = cursor.fetchone()

            if existing:
                if existing[0] >= insight.confidence_score:
                    print(f"[Fabric Consistency] ℹ️ Ignored update. Existing solution has higher or equal confidence ({existing[0]} >= {insight.confidence_score})")
                    return True # It's not a failure, just skipped
                else:
                    print(f"[Fabric Consistency] 🔄 Upgrading existing solution for '{insight.problem_signature}'")
                    cursor.execute("""
                        UPDATE insights SET solution=?, confidence_score=?, source_agent_id=?, metadata=?
                        WHERE problem_signature=?
                    """, (insight.solution, insight.confidence_score, insight.source_agent_id, json.dumps(insight.metadata), insight.problem_signature))
            else:
                cursor.execute("""
                    INSERT INTO insights (problem_signature, solution, confidence_score, source_agent_id, metadata)
                    VALUES (?, ?, ?, ?, ?)
                """, (insight.problem_signature, insight.solution, insight.confidence_score, insight.source_agent_id, json.dumps(insight.metadata)))
            
            self.conn.commit()
            print(f"[Fabric ACCEPT] ✅ Insight for '{insight.problem_signature}' safely committed to collective memory.")

        except Exception as e:
            print(f"[Fabric ERROR] ❌ Database error: {e}")
            return False

        # 3. Accelerator Mechanism (Broadcast)
        print(f"[Fabric BROADCAST] 📡 Broadcasting insight to {len(self.subscribers)} registered nodes...")
        for subscriber in self.subscribers:
            subscriber(insight)
        
        return True

    def query_insight(self, problem_signature: str) -> Optional[Insight]:
        """Query the Fabric for an existing verified solution."""
        cursor = self.conn.cursor()
        cursor.execute("SELECT problem_signature, solution, confidence_score, source_agent_id, metadata FROM insights WHERE problem_signature = ?", (problem_signature,))
        row = cursor.fetchone()
        if row:
            return Insight(row[0], row[1], row[2], row[3], json.loads(row[4]))
        return None

File: backend/src/fabric.ts
import sqlite3 from "sqlite3";
import { Insight } from "./models";
import { AutonomousGuardrail } from "./guardrail";

// Helper for Cosine Similarity
function cosineSimilarity(vecA: number[], vecB: number[]): number {
  if (vecA.length !== vecB.length) return 0;
  let dotProduct = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < vecA.length; i++) {
    dotProduct += vecA[i] * vecB[i];
    normA += vecA[i] * vecA[i];
    normB += vecB[i] * vecB[i];
  }
  if (normA === 0 || normB === 0) return 0;
  return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
}

export class CognitionFabric {
  private db: sqlite3.Database;
  private guardrail: AutonomousGuardrail;

  constructor() {
    this.db = new sqlite3.Database("./fabric_ledger.db");
    this.guardrail = new AutonomousGuardrail(0.8);
    this.initDb();
  }

  private initDb() {
    this.db.serialize(() => {
      this.db.run(`
        CREATE TABLE IF NOT EXISTS insights (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          problem_signature TEXT UNIQUE,
          solution TEXT,
          confidence_score REAL,
          source_agent_id TEXT,
          metadata TEXT,
          vector_embedding TEXT,
          timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
        )
      `);
    });
  }

  public publishInsight(insight: Insight, vector: number[], callback: (i: Insight) => void): {success: boolean, message: string} {
    const isSafe = this.guardrail.validate(insight);
    if (!isSafe.valid) {
      return { success: false, message: `[Guardrail BLOCK] 🛑 ${isSafe.reason}` };
    }

    const stmt = this.db.prepare(`
      INSERT INTO insights (problem_signature, solution, confidence_score, source_agent_id, metadata, vector_embedding)
      VALUES (?, ?, ?, ?, ?, ?)
      ON CONFLICT(problem_signature) DO UPDATE SET
        solution = excluded.solution,
        confidence_score = excluded.confidence_score,
        source_agent_id = excluded.source_agent_id,
        metadata = excluded.metadata,
        vector_embedding = excluded.vector_embedding,
        timestamp = CURRENT_TIMESTAMP
      WHERE excluded.confidence_score > insights.confidence_score
    `);

    const metaStr = JSON.stringify(insight.metadata || {});
    const vecStr = JSON.stringify(vector);

    stmt.run([
      insight.problem_signature,
      insight.solution,
      insight.confidence_score,
      insight.source_agent_id,
      metaStr,
      vecStr
    ], function(err) {
      if (err) console.error("SQLite Error:", err);
      // We only callback if rows were changed (meaning it wasn't ignored due to lower confidence)
      if (this.changes > 0) {
        callback(insight);
      }
    });

    return { success: true, message: "Insight committed to Fabric Ledger." };
  }

  public getAllInsights(callback: (insights: Insight[]) => void) {
    this.db.all("SELECT problem_signature, solution, confidence_score, source_agent_id, metadata, timestamp FROM insights ORDER BY id DESC", [], (err, rows: any[]) => {
      if (err) {
        callback([]);
        return;
      }
      callback(rows.map(r => ({
        ...r,
        metadata: JSON.parse(r.metadata)
      })));
    });
  }

  // SEMANTIC VECTOR SEARCH
  public findSemanticMatch(queryVector: number[], similarityThreshold: number = 0.85, callback: (match: Insight | null, similarity: number) => void) {
    this.db.all("SELECT problem_signature, solution, confidence_score, source_agent_id, metadata, vector_embedding FROM insights", [], (err, rows: any[]) => {
      if (err || !rows || rows.length === 0) {
        callback(null, 0);
        return;
      }

      let bestMatch: Insight | null = null;
      let highestSimilarity = -1;

      for (const row of rows) {
        if (!row.vector_embedding) continue;
        const storedVector = JSON.parse(row.vector_embedding);
        const sim = cosineSimilarity(queryVector, storedVector);
        
        if (sim > highestSimilarity) {
          highestSimilarity = sim;
          bestMatch = {
            problem_signature: row.problem_signature,
            solution: row.solution,
            confidence_score: row.confidence_score,
            source_agent_id: row.source_agent_id,
            metadata: JSON.parse(row.metadata)
          };
        }
      }

      if (highestSimilarity >= similarityThreshold && bestMatch) {
        callback(bestMatch, highestSimilarity);
      } else {
        callback(null, highestSimilarity);
      }
    });
  }

  // Wipes the database for hackathon demos
  public clearFabric(callback: () => void) {
    this.db.run("DELETE FROM insights", [], (err) => {
      if (err) console.error("Error clearing DB:", err);
      callback();
    });
  }
}
